examples/cis2-wccd-upgrade/src/lib.rs

```diff
@@ -0,0 +1,2030 @@
+            }))?;
+
+            // Only log a transfer event if receiver is not the one who payed for this.
+            if sender != receive_address {
+                logger.log(&Cis2Event::Transfer(TransferEvent {
+                    token_id: TOKEN_ID_WCCD,
+                    amount:   ContractTokenAmount::from(amount.micro_ccd),
+                    from:     sender,
+                    to:       receive_address,
+                }))?;
+            }
+
+            // If the receiver is a contract: invoke the receive hook function.
+            if let Receiver::Contract(address, function) = params.to {
+                let parameter = OnReceivingCis2Params {
+                    token_id: TOKEN_ID_WCCD,
+                    amount:   ContractTokenAmount::from(amount.micro_ccd),
+                    from:     sender,
+                    data:     params.data,
+                };
+                host.invoke_contract(
+                    &address,
+                    &parameter,
+                    function.as_entrypoint_name(),
+                    Amount::zero(),
+                )
+                .unwrap_abort();
+                Ok(())
+            } else {
+                Ok(())
+            }
+        }
+    }
+}
+
+/// Unwrap an amount of wCCD tokens into CCD
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "unwrap",
+    parameter = "UnwrapParams",
+    enable_logger,
+    mutable
+)]
+fn contract_unwrap<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    let slot_time = ctx.metadata().slot_time();
+
+    match host.state().state_address {
+        None => bail!(concordium_cis2::Cis2Error::Custom(OnlyState)),
+
+        Some(state_address) => {
+            let unpause_time = host
+                .invoke_contract_raw(
+                    &state_address,
+                    Parameter(&[]),
+                    EntrypointName::new_unchecked("get_unpause_time"),
+                    Amount::zero(),
+                )?
+                .1;
+
+            let unpause_time = if let Some(mut unpause_time) = unpause_time {
+                unpause_time.get()?
+            } else {
+                return Err(concordium_cis2::Cis2Error::Custom(StateInvokeError));
+            };
+
+            ensure!(slot_time >= unpause_time, concordium_cis2::Cis2Error::Custom(ContractPaused));
+            let params: UnwrapParams = ctx.parameter_cursor().get()?;
+            // Get the sender who invoked this contract function.
+            let sender = ctx.sender();
+            let state = host.state_mut();
+            ensure!(
+                sender == params.owner || state.is_operator(&sender, &params.owner),
+                ContractError::Unauthorized
+            );
+
+            // Update the state.
+            state.burn(&TOKEN_ID_WCCD, params.amount, &params.owner)?;
+
```

Comment:
Presumably this burning will be part of the state contract, right?

When that is the case, would it not make more sense that the state contract itself checks whether it is paused or not?

You save a call and some complexity.

Reply:
My plan is to keep all the logic (including this burn function and any pausing/unpausing function) in the implementation contract. That way we can upgrade the logic.